File: components/gmaps-flow/version-1.0/google-map-flow/src/main/resources/META-INF/resources/frontend/info-window.js
/** 
@license
Copyright 2010-2019 Neotropic SAS <contact@neotropic.co>.

Licensed under the Apache License, Version 2.0 (the "License"); 
you may not use this file except in compliance with the License. 
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software 
distributed under the License is distributed on an "AS IS" BASIS, 
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
See the License for the specific language governing permissions and 
limitations under the License.
*/
import {html, PolymerElement} from '@polymer/polymer/polymer-element.js';
import {FlattenedNodesObserver} from '@polymer/polymer/lib/utils/flattened-nodes-observer.js';
/**
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
class InfoWindow extends PolymerElement {
  static get is() {
    return 'info-window';
  }
  static get template() {
    return html`
      <style>
        :host {
          display: block;
        }
      </style>
    `;
  }
  static get properties() {
    return {
      disableAutoPan: {
        type: Boolean,
        observer: '_disableAutoPanChanged'
      },
      maxWidth: {
        type: Number,
        observer: '_maxWidthChanged'
      },
      pixelOffset: {
        type: Object,
        observer: '_pixelOffsetChanged'
      },
      position: {
        type: Object,
        observer: '_positionChanged'
      },
      zIndex: {
        type: Number,
        observer: '_zIndexChanged'
      }
    };
  }
  /**
   * @return {google.maps.MVCObject} The google.maps.InfoWindow
   */
  getMVCObject() {
    return this.infoWindow;
  }
  added() {
    this.infoWindow = new google.maps.InfoWindow();
    google.maps.event.addListener(this.infoWindow, 'closeclick', () => {
      this.dispatchEvent(new CustomEvent('info-window-close-click'));
    });
    google.maps.event.addListener(this.infoWindow, 'content_changed', () => {
      this.dispatchEvent(new CustomEvent('info-window-content-changed'));
    });
    google.maps.event.addListener(this.infoWindow, 'domready', () => {
      this.dispatchEvent(new CustomEvent('info-window-dom-ready'));
    });
    google.maps.event.addListener(this.infoWindow, 'position_changed', () => {
      this.dispatchEvent(new CustomEvent('info-window-position-changed'));
    });
    google.maps.event.addListener(this.infoWindow, 'zindex_changed', () => {
      this.dispatchEvent(new CustomEvent('info-window-zindex-changed'));
    });
    this._observer = new FlattenedNodesObserver(this, info => {
      this._addedNodes(info.addedNodes);
      this._removedNodes(info.removedNodes);
    });
    this.dispatchEvent(new CustomEvent('info-window-added'));
  }
  _addedNodes(addedNodes) {
    addedNodes.forEach(value => {
      this._contentChanged(value);
    });
  }
  _removedNodes(removedNodes) {
  }
  removed() {
    this.close();
  }
  close() {
    if (this.infoWindow)
      this.infoWindow.close();
  }
  /**
   * 
   * @param {google.maps.Map} map 
   * @param {google.maps.MVCObject} anchor 
   */
  open(map, anchor) {
    if (this.infoWindow) {
      if (map && anchor)
        this.infoWindow.open(map, anchor);
      else if (map)
        this.infoWindow.open(map);
      else if (anchor)
        this.infoWindow.open(anchor);
      else
        this.infoWindow.open();
    } else {
      
    }
  }
  /**
   * 
   * @param {*} content 
   */
  _contentChanged(content) {
    if (this.infoWindow)
      this.infoWindow.setOptions({content: content});
  }
  /**
   * 
   * @param {boolean} newValue 
   * @param {*} oldValue 
   */
  _disableAutoPanChanged(newValue, oldValue) {
    if (this.infoWindow)
      this.infoWindow.setOptions({disableAutoPan: newValue});
  }
  /**
   * 
   * @param {number} newValue 
   * @param {*} oldValue 
   */
  _maxWidthChanged(newValue, oldValue) {
    if (this.infoWindow)
      this.infoWindow.setOptions({maxWidth: newValue});
  }
  /**
   * 
   * @param {object} newValue {width: 0, height: 0}
   * @param {*} oldVlue 
   */
  _pixelOffsetChanged(newValue, oldValue) {
    if (this.infoWindow)
      this.infoWindow.setOptions({pixelOffset: new google.map.Size(newValue.width, newValue.height)});
  }
  /**
   * 
   * @param {object} newValue {lat: 0, lng: 0}
   * @param {*} oldValue 
   */
  _positionChanged(newValue, oldValue) {
    if (this.infoWindow)
      this.infoWindow.setOptions({position: newValue});
  }
  /**
   * 
   * @param {number} newValue 
   * @param {*} oldValue 
   */
  _zIndexChanged(newValue, oldValue) {
    if (this.infoWindow)
      this.infoWindow.setOptions({zIndex: newValue});
  }
}
window.customElements.define(InfoWindow.is, InfoWindow);

File: components/gmaps-flow/version-1.0/google-map-flow/src/main/java/com/neotropic/flow/component/googlemap/InfoWindow.java
/*
 *  Copyright 2010-2020 Neotropic SAS <contact@neotropic.co>.
 *  
 *  Licensed under the Apache License, Version 2.0 (the "License"); 
 *  you may not use this file except in compliance with the License. 
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software 
 *  distributed under the License is distributed on an "AS IS" BASIS, 
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *  See the License for the specific language governing permissions and 
 *  limitations under the License.
 */
package com.neotropic.flow.component.googlemap;

import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.HasComponents;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.shared.Registration;
import elemental.json.Json;
import elemental.json.JsonNull;
import elemental.json.JsonObject;
import java.io.Serializable;

/**
 *
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
@Tag("info-window")
@JsModule("./info-window.js")
public class InfoWindow extends Component implements HasComponents {
    public InfoWindow() {
    }
    public boolean getDisableAutoPan() {
        return getElement().getProperty(Constants.Property.DISABLE_AUTO_PAN, false);
    }
    public void setDisableAutoPan(boolean disableAutoPan) {
        getElement().setProperty(Constants.Property.DISABLE_AUTO_PAN, disableAutoPan);
    }
    public double getMaxWidth() {
        return getElement().getProperty(Constants.Property.MAX_WIDTH, 0);
    }
    public void setMaxWidth(double maxWidth) {
        getElement().setProperty(Constants.Property.MAX_WIDTH, maxWidth);
    }
    public Size getPixelOffset() {
        Serializable pixelOffset = getElement().getPropertyRaw(Constants.Property.PIXEL_OFFSET);
        if (pixelOffset instanceof JsonNull)
            return null;
        return new Size((JsonObject) pixelOffset);
    }
    public void setPixelOffset(Size pixelOffset) {
        if (pixelOffset != null)
            getElement().setPropertyJson(Constants.Property.PIXEL_OFFSET, pixelOffset.toJson());
        else
            getElement().setPropertyJson(Constants.Property.PIXEL_OFFSET, Json.createNull());
        
    }
    public LatLng getPosition() {
        Serializable position = getElement().getPropertyRaw(Constants.Property.POSITION);
        if (position instanceof JsonNull)
            return null;
        return new LatLng((JsonObject) position);
    }
    public void setPosition(LatLng position) {
        if (position != null)
            getElement().setPropertyJson(Constants.Property.POSITION, position.toJson());
        else
            getElement().setPropertyJson(Constants.Property.POSITION, Json.createNull());
    }
    public double getZIndex() {
        return getElement().getProperty(Constants.Property.Z_INDEX, 0);
    }
    public void setZIndex(double zIndex) {
        getElement().setProperty(Constants.Property.Z_INDEX, zIndex);
    }
    public void open(GoogleMap map, Component component) {
        getElement().executeJs(
            "this.open($0.getMVCObject(), $1.getMVCObject())", 
            map, component);
    }
    public void open(GoogleMap map) {
        getElement().executeJs(
            "this.open($0.getMVCObject())", 
            map);
    }
    public void close() {
        getElement().executeJs("this.close()");
    }
    public Registration addInfoWindowCloseClickListener(ComponentEventListener<GoogleMapEvent.InfoWindowCloseClickEvent> listener) {
        return addListener(GoogleMapEvent.InfoWindowCloseClickEvent.class, listener);
    }
    public Registration addInfoWindowContentChangedListener(ComponentEventListener<GoogleMapEvent.InfoWindowContentChangedEvent> listener) {
        return addListener(GoogleMapEvent.InfoWindowContentChangedEvent.class, listener);
    }
    public Registration addInfoWindowDomReadyListener(ComponentEventListener<GoogleMapEvent.InfoWindowDomReadyEvent> listener) {
        return addListener(GoogleMapEvent.InfoWindowDomReadyEvent.class, listener);
    }
    public Registration addInfoWindowPositionChangedListener(ComponentEventListener<GoogleMapEvent.InfoWindowPositionChangedEvent> listener) {
        return addListener(GoogleMapEvent.InfoWindowPositionChangedEvent.class, listener);
    }
    public Registration addInfoWindowZIndexChangedListener(ComponentEventListener<GoogleMapEvent.InfoWindowZIndexChangedEvent> listener) {
        return addListener(GoogleMapEvent.InfoWindowZIndexChangedEvent.class, listener);
    }
    public Registration addInfoWindowAddedListener(ComponentEventListener<GoogleMapEvent.InfoWindowAddedEvent> listener) {
        return addListener(GoogleMapEvent.InfoWindowAddedEvent.class, listener);
    }
}